src/dbus/dbus_message_helper_openthread.cpp

```diff
@@ -0,0 +1,366 @@
+    auto            args     = std::tie(aValue.mPrefix, preference);
+    auto            flagArgs = std::tie(preferred, slaac, dhcp, configure, defaultRoute, onMesh, stable);
+    otbrError       err      = OTBR_ERROR_NONE;
+
+    VerifyOrExit(dbus_message_iter_get_arg_type(aIter) == DBUS_TYPE_STRUCT, err = OTBR_ERROR_DBUS);
+    dbus_message_iter_recurse(aIter, &sub);
+    SuccessOrExit(err = ConvertToTuple(&sub, args));
+    VerifyOrExit(dbus_message_iter_get_arg_type(&sub) == DBUS_TYPE_STRUCT, err = OTBR_ERROR_DBUS);
+    dbus_message_iter_recurse(&sub, &flagsSub);
+    SuccessOrExit(err = ConvertToTuple(&flagsSub, flagArgs));
+
+    aValue.mPreference   = preference;
+    aValue.mPreferred    = preferred;
+    aValue.mSlaac        = slaac;
+    aValue.mDhcp         = dhcp;
+    aValue.mConfigure    = configure;
+    aValue.mDefaultRoute = defaultRoute;
+    aValue.mOnMesh       = onMesh;
+    aValue.mStable       = stable;
+exit:
+    return err;
+}
+
+otbrError DBusMessageEncode(DBusMessageIter *aIter, const otMeshLocalPrefix &aValue)
+{
+    std::vector<uint8_t> data(&aValue.m8[0], &aValue.m8[0] + sizeof(aValue.m8));
+
+    return DBusMessageEncode(aIter, data);
+}
+
+otbrError DBusMessageExtract(DBusMessageIter *aIter, otMeshLocalPrefix &aValue)
+{
+    std::vector<uint8_t> data;
+    otbrError            err = OTBR_ERROR_NONE;
+
+    SuccessOrExit(err = DBusMessageExtract(aIter, data));
+    VerifyOrExit(data.size() == sizeof(aValue.m8), err = OTBR_ERROR_NONE);
+    memcpy(aValue.m8, &data[0], data.size());
+exit:
+    return err;
+}
+
+otbrError DBusMessageEncode(DBusMessageIter *aIter, const otMacCounters &aValue)
+{
+    auto args = std::tie(
+        aValue.mTxTotal, aValue.mTxUnicast, aValue.mTxBroadcast, aValue.mTxAckRequested, aValue.mTxAcked,
+        aValue.mTxNoAckRequested, aValue.mTxData, aValue.mTxDataPoll, aValue.mTxBeacon, aValue.mTxBeaconRequest,
+        aValue.mTxOther, aValue.mTxRetry, aValue.mTxErrCca, aValue.mTxErrAbort, aValue.mTxErrBusyChannel,
+        aValue.mRxTotal, aValue.mRxUnicast, aValue.mRxBroadcast, aValue.mRxData, aValue.mRxDataPoll, aValue.mRxBeacon,
+        aValue.mRxBeaconRequest, aValue.mRxOther, aValue.mRxAddressFiltered, aValue.mRxDestAddrFiltered,
+        aValue.mRxDuplicated, aValue.mRxErrNoFrame, aValue.mRxErrUnknownNeighbor, aValue.mRxErrInvalidSrcAddr,
+        aValue.mRxErrSec, aValue.mRxErrFcs, aValue.mRxErrOther);
+    DBusMessageIter sub;
+    otbrError       err = OTBR_ERROR_NONE;
+
+    VerifyOrExit(dbus_message_iter_open_container(aIter, DBUS_TYPE_STRUCT, nullptr, &sub));
+    SuccessOrExit(err = ConvertToDBusMessage(&sub, args));
+    VerifyOrExit(dbus_message_iter_close_container(aIter, &sub) == true, err = OTBR_ERROR_DBUS);
+exit:
+    return err;
+}
+
+otbrError DBusMessageExtract(DBusMessageIter *aIter, otMacCounters &aValue)
+{
+    auto args = std::tie(
+        aValue.mTxTotal, aValue.mTxUnicast, aValue.mTxBroadcast, aValue.mTxAckRequested, aValue.mTxAcked,
```

Comment:
defines a static tuple or macro to elimit copying.

Reply:
std::tie. Constructs a tuple object whose elements are **references** to the arguments in args , in the same order. This allows a set of objects to act as a tuple, which is especially useful to unpack tuple objects.

So we are not copying these arguments.

Reply:
@gjc13 sorry for the unclarity, `copying` I mean, is copying of the code. Since this argument list is significant long. I think we should define a macro/function to generate the arguments tuple. Especially if the order of each item matters.